solr/modules/monitor/src/test/org/apache/solr/monitor/MonitorSolrQueryTest.java

```diff
@@ -0,0 +1,643 @@
+          writeToDocList,
+          QUERY_MATCH_TYPE_KEY,
+          "simple"
+        };
+
+    QueryResponse response = query(params);
+    if (writeToDocList) {
+      assertEquals(1, ((SolrDocumentList) response.getResponse().get("response")).size());
+    }
+    validate(response, 0, List.of("0"), writeToDocList);
+
+    del(MonitorFields.QUERY_ID + ":0");
+    commit();
+    response = query(CommonParams.Q, "*:*");
+    assertEquals(0, ((SolrDocumentList) response.getResponse().get("response")).size());
+  }
+
+  static Map<Object, Object> queryMatch(
+      String queryId,
+      String field1,
+      List<HighlightsMatch.Hit> firstFieldHits,
+      String field2,
+      List<HighlightsMatch.Hit> secondFieldHits) {
+    Map<Object, Object> queryMatch = new LinkedHashMap<>();
+    queryMatch.put(MonitorFields.QUERY_ID, queryId);
+    Map<Object, Object> matchHits = new LinkedHashMap<>();
+    var outHits1 = map(firstFieldHits);
+    if (!outHits1.isEmpty()) {
+      matchHits.put(field1, outHits1);
+    }
+    var outHits2 = map(secondFieldHits);
+    if (!outHits2.isEmpty()) {
+      matchHits.put(field2, outHits2);
+    }
+    queryMatch.put(MonitorConstants.HITS_KEY, matchHits);
+    return queryMatch;
+  }
+
+  static List<Object> map(List<HighlightsMatch.Hit> hits) {
+    List<Object> outHits = new ArrayList<>();
+    for (var hit : hits) {
+      Map<Object, Object> hit00 = new LinkedHashMap<>();
+      hit00.put("startPosition", hit.startPosition);
+      hit00.put("endPosition", hit.endPosition);
+      hit00.put("startOffset", hit.startOffset);
+      hit00.put("endOffset", hit.endOffset);
+      outHits.add(hit00);
+    }
+    return outHits;
+  }
+
+  void validate(QueryResponse response, int doc, Object expectedValue, boolean writeToDocList) {
+    var monitorQueries = monitorQueries(response, doc);
+    assertEquals(expectedValue, monitorQueries);
+    if (writeToDocList) {
+      validateDocList(response, doc, expectedValue);
+    }
+  }
+
+  void validate(
+      QueryResponse response, int doc, int query, Object expectedValue, boolean writeToDocList) {
+    var monitorQueries = monitorQueries(response, doc);
+    assertTrue(monitorQueries.size() > query);
+    assertEquals(expectedValue, monitorQueries.get(query));
+    if (writeToDocList) {
+      validateDocList(response, doc, expectedValue);
+    }
+  }
+
+  void validateDocList(QueryResponse response, int doc, Object expectedValue) {
+    SolrDocumentList actualValues = (SolrDocumentList) response.getResponse().get("response");
+    // minimal checks only so far; TODO: make this more comprehensive
+    if (expectedValue instanceof List) {
+      List<Object> expectedValues = (List) expectedValue;
+      if (expectedValues.size() == 1 && actualValues.size() <= 2) {
+        assertEquals(
+            expectedValues.get(0),
+            actualValues.get(actualValues.size() - 1 - doc).getFieldValue(MonitorFields.QUERY_ID));
+      }
+    }
+  }
```

Comment:
https://github.com/kotman12/solr/pull/4 tries to make the MonitorSolrQueryTest.validate's expectedValue argument type more specific (String of List<String> instead of Object) but even with that I haven't yet managed to create a `validateDocList` implementation for the existing test expectations. Not yet anyhow.